### PR TITLE
[INLONG-9314][Manager] Support cluster switch for InlongGroup

### LINKED ISSUE #9314: [Feature][Manager] Support cluster switch for InlongGroup
### Description

For better resource utilization, sometimes we need to move one group from one cluster to another.

The whole process can be divided into the following steps:
1. Check if the target cluster tag exists in the tenant, if not, bind them together.
2. Check if the current related sort clusters have the target cluster tag, if not, the switch failed.
3. Add backup_cluster_tag into the inlong_group_ext.
4. Replace the cluster with the backup_cluster_tag, and remove backup_cluster_tag.

### Use case

_No response_

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by vernedeng**
### Prepare a Pull Request
*(Change the title refer to the following example)*

- Title Example: [INLONG-XYZ][Component] Title of the pull request

*(The following *XYZ* should be replaced by the actual [GitHub Issue](https://github.com/apache/inlong/issues) number)*

- Fixes #9314 

### Motivation

*Explain here the context, and why you're making that change. What is the problem you're trying to solve?*

### Modifications

*Describe the modifications you've done.*

### Verifying this change

*(Please pick either of the following options)*

- [ ] This change is a trivial rework/code cleanup without any test coverage.

- [ ] This change is already covered by existing tests, such as:
  *(please describe tests)*

- [ ] This change added tests and can be verified as follows:

  *(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a follow-up issue for adding the documentation

### CODE REVIEW

File: inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java [hunk 1]
```diff
@@ -672,4 +697,134 @@ private void chkUnmodifiableParams(InlongGroupEntity entity, InlongGroupRequest
...
+            LOGGER.error("tenant cluster not found for tenant={}, cluster={}", groupEntity.getTenant(), clusterTag);
+            throw new BusinessException(ErrorCodeEnum.TENANT_CLUSTER_TAG_NOT_FOUND);
+        }
+
+        // check if all sink related sort cluster has the target cluster tag
+        List<StreamSink> sinks = streamSinkService.listSink(groupEntity.getInlongGroupId(), null);
+        for (StreamSink sink : sinks) {
+            String clusterName = sink.getInlongClusterName();
+            InlongClusterEntity clusterEntity =
+                    clusterEntityMapper.selectByNameAndType(clusterName, null);
```
Comment (fuweng11) [inline inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java:739]: Suggest use` clusterEntityMapper.selectByKey`.

File: inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java [hunk 2]
```diff
@@ -672,4 +697,134 @@ private void chkUnmodifiableParams(InlongGroupEntity entity, InlongGroupRequest
...
+        // trigger group workflow to rebuild configs
+        this.triggerWorkFlow(groupInfo, userInfo);
+        LOGGER.info("success to switch cluster tag for group={}, target tag={}", groupId, clusterTag);
+        return true;
+    }
+
+    @Override
+    @Transactional(rollbackFor = Throwable.class, isolation = Isolation.REPEATABLE_READ, propagation = Propagation.REQUIRES_NEW)
+    public Boolean finishTagSwitch(String groupId) {
+        LOGGER.info("start to finish switch cluster tag for group={}", groupId);
```
Comment (fuweng11) [inline inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java:775]: finish?

File: inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java [hunk 3]
```diff
@@ -672,4 +697,134 @@ private void chkUnmodifiableParams(InlongGroupEntity entity, InlongGroupRequest
...
+
+        InlongGroupInfo groupInfo = this.get(groupId);
```
Comment (fuweng11) [inline inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java:777]: Please add a status check for the group to ensure that the switch has been completed.

Comment (vernedeng) [reply inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java:777]: fixed, thx

File: inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupService.java [hunk 4]
```diff
@@ -207,4 +207,8 @@ void updateAfterApprove(
      */
     Map<String, Object> detail(String groupId);
 
+    Boolean startTagSwitch(String groupId, String clusterTag);
```
Comment (fuweng11) [inline inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupService.java:210]: Add doc.

File: inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java [hunk 5]
```diff
@@ -672,4 +699,150 @@ private void chkUnmodifiableParams(InlongGroupEntity entity, InlongGroupRequest
                 String.format("record has expired with record version=%d, request version=%d",
                         entity.getVersion(), request.getVersion()));
     }
+
+    @Override
+    @Transactional(rollbackFor = Throwable.class, isolation = Isolation.REPEATABLE_READ, propagation = Propagation.REQUIRES_NEW)
+    public Boolean startTagSwitch(String groupId, String clusterTag) {
+        LOGGER.info("start to switch cluster tag for group={}, target tag={}", groupId, clusterTag);
```
Comment (fuweng11) [inline inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java:706]: start to switch cluster tag for groupId=

File: inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java [hunk 6]
```diff
@@ -672,4 +699,150 @@ private void chkUnmodifiableParams(InlongGroupEntity entity, InlongGroupRequest
...
+
+        InlongGroupInfo groupInfo = this.get(groupId);
+
+        // check if the group mode is data sync mode
+        if (groupInfo.getInlongGroupMode() == 1) {
```
Comment (fuweng11) [inline inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java:711]: InlongConstants.DATASYNC_MODE.equals(groupEntity.getInlongGroupMode())

File: inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java [hunk 7]
```diff
@@ -672,4 +699,150 @@ private void chkUnmodifiableParams(InlongGroupEntity entity, InlongGroupRequest
...
+        }
+
+        // check if the group is under switching
+        List<InlongGroupExtInfo> groupExt = groupInfo.getExtList();
+        Set<String> keys = groupExt.stream()
+                .map(InlongGroupExtInfo::getKeyName)
+                .collect(Collectors.toSet());
+
+        if (keys.contains(BACKUP_CLUSTER_TAG) || keys.contains(BACKUP_MQ_RESOURCE)) {
+            String errMsg = String.format("switch failed, current group is under switching, group=[%s]", groupId);
```
Comment (fuweng11) [inline inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java:724]: groupId=

File: inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java [hunk 8]
```diff
@@ -672,4 +699,150 @@ private void chkUnmodifiableParams(InlongGroupEntity entity, InlongGroupRequest
...
+    @Override
+    @Transactional(rollbackFor = Throwable.class, isolation = Isolation.REPEATABLE_READ, propagation = Propagation.REQUIRES_NEW)
+    public Boolean startTagSwitch(String groupId, String clusterTag) {
+        LOGGER.info("start to switch cluster tag for group={}, target tag={}", groupId, clusterTag);
+
+        InlongGroupInfo groupInfo = this.get(groupId);
+
+        // check if the group mode is data sync mode
+        if (groupInfo.getInlongGroupMode() == 1) {
+            String errMSg = String.format("no need to switch sync mode group = {}", groupId);
```
Comment (fuweng11) [inline inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java:712]: groupId = {}

File: inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java [hunk 9]
```diff
@@ -672,4 +699,150 @@ private void chkUnmodifiableParams(InlongGroupEntity entity, InlongGroupRequest
...
+        InlongGroupEntity groupEntity = groupMapper.selectByGroupId(groupId);
+        if (groupEntity == null) {
+            LOGGER.error("inlong group not found by groupId={}", groupId);
+            throw new BusinessException(ErrorCodeEnum.GROUP_NOT_FOUND);
+        }
+
+        TenantClusterTagEntity tenantClusterTag =
+                tenantClusterTagMapper.selectByUniqueKey(clusterTag, groupEntity.getTenant());
+        if (tenantClusterTag == null) {
+            LOGGER.error("tenant cluster not found for tenant={}, cluster={}", groupEntity.getTenant(), clusterTag);
```
Comment (fuweng11) [inline inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java:739]: clusterTag=

File: inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java [hunk 10]
```diff
@@ -672,4 +699,150 @@ private void chkUnmodifiableParams(InlongGroupEntity entity, InlongGroupRequest
...
+
+        // check if all sink related sort cluster has the target cluster tag
+        List<StreamSink> sinks = streamSinkService.listSink(groupEntity.getInlongGroupId(), null);
+        for (StreamSink sink : sinks) {
+            String clusterName = sink.getInlongClusterName();
+            InlongClusterEntity clusterEntity =
+                    clusterEntityMapper.selectByNameAndType(clusterName,
+                            SinkType.relatedSortClusterType(sink.getSinkType()));
+            if (clusterEntity == null) {
+                String errMsg = String.format("find no cluster with name=[%s]", clusterName);
```
Comment (fuweng11) [inline inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java:751]: clusterName=[%s]

File: inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java [hunk 11]
```diff
@@ -672,4 +699,150 @@ private void chkUnmodifiableParams(InlongGroupEntity entity, InlongGroupRequest
...
+        request.setInlongClusterTag(clusterTag);
+        request.getExtList().add(new InlongGroupExtInfo(null, groupId, BACKUP_CLUSTER_TAG, oldClusterTag));
+        request.getExtList().add(new InlongGroupExtInfo(null, groupId, BACKUP_MQ_RESOURCE, request.getMqResource()));
+        request.getExtList().add(new InlongGroupExtInfo(null, groupId, CLUSTER_SWITCH_TIME,
+                LocalDateTime.now().toString()));
+        this.update(request, userInfo.getName());
+
+        // trigger group workflow to rebuild configs
+        this.triggerWorkFlow(groupInfo, userInfo);
+        LOGGER.info("success to switch cluster tag for group={}, target tag={}", groupId, clusterTag);
```
Comment (fuweng11) [inline inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java:778]: groupId={}

File: inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java [hunk 12]
```diff
@@ -672,4 +699,150 @@ private void chkUnmodifiableParams(InlongGroupEntity entity, InlongGroupRequest
...
+        return true;
+    }
+
+    @Override
+    @Transactional(rollbackFor = Throwable.class, isolation = Isolation.REPEATABLE_READ, propagation = Propagation.REQUIRES_NEW)
+    public Boolean finishTagSwitch(String groupId) {
+        LOGGER.info("start to finish switch cluster tag for group={}", groupId);
```
Comment (fuweng11) [inline inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java:785]: groupId={}

File: inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java [hunk 13]
```diff
@@ -672,4 +699,150 @@ private void chkUnmodifiableParams(InlongGroupEntity entity, InlongGroupRequest
...
+            throw new BusinessException(ErrorCodeEnum.GROUP_UPDATE_NOT_ALLOWED, errMsg);
+        }
+
+        // check if the group is under switching
+        List<InlongGroupExtInfo> groupExt = groupInfo.getExtList();
+        Map<String, InlongGroupExtInfo> extInfoMap = groupExt.stream()
+                .collect(Collectors.toMap(InlongGroupExtInfo::getKeyName, v -> v));
+
+        if (!extInfoMap.containsKey(BACKUP_CLUSTER_TAG) || !extInfoMap.containsKey(BACKUP_MQ_RESOURCE)) {
+            String errMsg = String.format("finish switch failed, current group is not under switching, group=[%s]",
```
Comment (fuweng11) [inline inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java:804]: Ditto.

File: inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java (hunk 6 above)
Comment (vernedeng) [reply inlong-manager/manager-service/src/main/java/org/apache/inlong/manager/service/group/InlongGroupServiceImpl.java:711]: fixed, thx

**Review by dockerzhang — APPROVED** [review]

**Pull request merged**